// @sentio/sdk/lib/core/handler-register.js
import { ServerError, Status } from 'nice-grpc';
export class HandlerRegister {
    handlerByChain = new Map();
    /**
     * Register a handler function with chain ID and handle type
     * @returns handler ID
     */
    register(handler, chainId) {
        const entries = this.handlerByChain.get(chainId) || [];
        const id = entries.length;
        const entry = {
            id,
            handler,
            chainId
        };
        entries.push(entry);
        this.handlerByChain.set(chainId, entries);
        return id;
    }
    /**
     * Get handler function by ID
     */
    getHandlerById(chainId, id) {
        const entries = this.handlerByChain.get(chainId);
        if (!entries || id < 0 || id >= entries.length) {
            throw new ServerError(Status.INTERNAL, `Handler with ID ${id} not found.`);
        }
        return entries[id].handler;
    }
    clear(chainId) {
        if (chainId) {
            this.handlerByChain.delete(chainId);
        }
        else {
            this.handlerByChain.clear();
        }
    }
}
