// @sentio/api/dist/src/sdk.gen.js
"use strict";
// This file is auto-generated by @hey-api/openapi-ts
Object.defineProperty(exports, "__esModule", { value: true });
exports.ForksService = exports.DebugAndSimulationService = exports.ProcessorService = exports.PriceService = exports.WebService = exports.DataService = exports.AlertsService = void 0;
const client_gen_1 = require("./client.gen");
class AlertsService {
    /**
     * Save an alert rule
     */
    static saveAlertRule(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/alerts/rule',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * List all alert rules for a project
     */
    static getAlertRules(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/alerts/rule/project/{projectId}',
            ...options
        });
    }
    /**
     * Delete an alert rule
     */
    static deleteAlertRule(options) {
        return (options.client ?? client_gen_1.client).delete({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/alerts/rule/{id}',
            ...options
        });
    }
    /**
     * Save an alert rule
     */
    static saveAlertRule2(options) {
        return (options.client ?? client_gen_1.client).put({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/alerts/rule/{id}',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Find an alert rule by id, and list all alerts for this rule
     */
    static getAlert(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/alerts/{ruleId}',
            ...options
        });
    }
}
exports.AlertsService = AlertsService;
class DataService {
    /**
     * Cancel SQL Query
     * Cancel a SQL query by execution_id.
     */
    static cancelSqlQuery(options) {
        return (options.client ?? client_gen_1.client).put({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/cancel_query/{executionId}',
            ...options
        });
    }
    /**
     * Execute SQL
     * Execute SQL in a project. Go to "Data Studio" -> "SQL Editor", write your query and then click "Export as cURL"
     *
     * ![screenshot](https://raw.githubusercontent.com/sentioxyz/docs/main/.gitbook/assets/image%20(102).png)
     *
     * Find more: https://docs.sentio.xyz/reference/data#sql-api
     */
    static executeSql(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/execute',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Execute SQL by Async
     * Execute SQL in a project asynchronously.
     */
    static executeSqlAsync(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/execute/async',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Query SQL Execution Detail
     * Query the execution detail of a SQL query by execution_id.
     */
    static querySqlExecutionDetail(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/query_execution_detail/{executionId}',
            ...options
        });
    }
    /**
     * Query SQL Result
     * Query the result of a SQL query by execution_id.
     */
    static querySqlResult(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/query_result/{executionId}',
            ...options
        });
    }
    /**
     * Save SQL
     * Save or update a SQL query in a project.
     */
    static saveSql(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/save',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Save SQL
     * Save or update a SQL query in a project.
     */
    static saveSql2(options) {
        return (options.client ?? client_gen_1.client).put({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/analytics/{owner}/{slug}/sql/save',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Query event logs
     */
    static queryLog(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/eventlogs/{owner}/{slug}',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Query event logs
     */
    static queryLog2(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/eventlogs/{owner}/{slug}/query',
            ...options
        });
    }
    /**
     * List coins
     * Get a list of coins in a project.
     */
    static listCoins2(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/insights/coins',
            ...options
        });
    }
    /**
     * List coins
     * Get a list of coins in a project.
     */
    static listCoins(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/insights/{owner}/{slug}/coins',
            ...options
        });
    }
    /**
     * Insight Query
     * Query for metrics,event logs and coin prices in a project.
     */
    static query(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/insights/{owner}/{slug}/query',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get a list of metrics in a project
     */
    static getMetrics(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/metrics',
            ...options
        });
    }
    /**
     * Metric instant queries
     */
    static queryInstant(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/metrics/{owner}/{slug}/query',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Metric range queries
     * The easiest way to build query is through UI, you could first create an insight chart, and then **Export as cURL**.
     *
     * ![screenshot](https://raw.githubusercontent.com/sentioxyz/docs/main/.gitbook/assets/image%20(101).png)
     */
    static queryRange(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/metrics/{owner}/{slug}/query_range',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
}
exports.DataService = DataService;
class WebService {
    /**
     * List all dashboards in a project
     */
    static listDashboards(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/dashboards',
            ...options
        });
    }
    /**
     * Import a dashboard to another dashboard
     */
    static importDashboard(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/dashboards/json',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Delete a dashboard by id
     */
    static deleteDashboard(options) {
        return (options.client ?? client_gen_1.client).delete({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/dashboards/{dashboardId}',
            ...options
        });
    }
    /**
     * Get a dashboard by id
     */
    static getDashboard(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/dashboards/{dashboardId}',
            ...options
        });
    }
    /**
     * Export a dashboard to json
     */
    static exportDashboard(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/dashboards/{dashboardId}/json',
            ...options
        });
    }
    /**
     * Get project details
     */
    static getProject(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/project/{owner}/{slug}',
            ...options
        });
    }
    /**
     * Get project details
     */
    static getProjectById(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/project/{projectId}',
            ...options
        });
    }
    /**
     * Get project list
     */
    static getProjectList(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/projects',
            ...options
        });
    }
    /**
     * List all dashboards in a project
     */
    static listDashboards2(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/projects/{owner}/{slug}/dashboards',
            ...options
        });
    }
    /**
     * Get a dashboard by id
     */
    static getDashboard2(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/projects/{owner}/{slug}/dashboards/{dashboardId}',
            ...options
        });
    }
}
exports.WebService = WebService;
class PriceService {
    /**
     * Get price
     * GetPrice returns the price of a given coin identifier, in a best effort way.
     * If we do not have any price data for the given coin, we will return NOT_FOUND error.
     * If we have at least one price data for the given coin, we will return it with the actual timestamp.
     * Client is responsible for checking the timestamp and decide whether to use the price or not.
     */
    static getPrice(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/prices',
            ...options
        });
    }
    /**
     * Add coin by Gecko
     * adds a coin by its coingecko id.
     *
     * coingecko id the API ID of the coin in coingecko web page.
     *
     * please AWARE that the coingecko id is NOT the same as the symbol of the coin.
     *
     * ![screenshot](https://raw.githubusercontent.com/sentioxyz/docs/main/.gitbook/assets/coingecko_apiid.png)
     */
    static addCoinByGecko(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/prices/add_coin_by_gecko',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Batch get prices
     */
    static batchGetPrices(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/prices/batch',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Check latest price
     */
    static checkLatestPrice(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/prices/check_latest',
            ...options
        });
    }
    /**
     * List coins
     */
    static priceListCoins(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/prices/coins',
            ...options
        });
    }
}
exports.PriceService = PriceService;
class ProcessorService {
    /**
     * Get processor status
     */
    static getProcessorStatusV2(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/processors/{owner}/{slug}/status',
            ...options
        });
    }
}
exports.ProcessorService = ProcessorService;
class DebugAndSimulationService {
    /**
     * Estimate gas price
     */
    static getEstimatedGasPrice(options) {
        return (options?.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/estimated_gas_price',
            ...options
        });
    }
    /**
     * Search transactions
     */
    static searchTransactions(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/search_transactions',
            ...options
        });
    }
    /**
     * Get list of simulations
     */
    static getSimulations(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/simulation',
            ...options
        });
    }
    /**
     * Get simulation by ID
     */
    static getSimulation(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/simulation/{simulationId}',
            ...options
        });
    }
    /**
     * Get bundle simulation by ID
     */
    static getSimulationBundleInProject(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/simulation_bundle/{bundleId}',
            ...options
        });
    }
    /**
     * Get trace by bundle simulation
     */
    static getCallTraceByBundle(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/{chainId}/bundle/{bundleId}/call_trace',
            ...options
        });
    }
    /**
     * Run simulation
     * Create a new transaction simulation. The simulation body should be included in the request body.
     * Your simulations will be saved, and a unique ID for each simulation is included in the response. It will be useful for fetching simulation details.
     */
    static simulateTransaction(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/{chainId}/simulation',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get trace by simulation
     */
    static getCallTraceBySimulation(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/{chainId}/simulation/{simulationId}/call_trace',
            ...options
        });
    }
    /**
     * Run bundle simulation
     * You could also create bundle simulations so that one transaction could be executed one after another. For `blockNumber` `transactionIndex` `networkId` `stateOverrides` and `blockOverrides` fields, only the first simulation takes effect.
     */
    static simulateTransactionBundle(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/{chainId}/simulation_bundle',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get trace by transaction
     * API to get Sentio call trace. It takes `txId.txHash` and `chainSpec.chainId` arguments, where the first is transaction hash, and the second is the numeric ethereum chain ID.
     *
     * The results looks very similar to the normal [Ethereum call trace](https://raw.githubusercontent.com/sentioxyz/docs/main/.gitbook/assets/image%20(2)%20(1)%20(1)%20(1).png). But we have an additional `startIndex` and `startIndex` on each trace entry even for the LOG, representing the execution order in the trace.
     *
     * This allows you to build chart that marks the order of fund flow.
     *
     * ![screenshot](https://raw.githubusercontent.com/sentioxyz/docs/main/.gitbook/assets/image%20(2)%20(1)%20(1)%20(1).png)
     */
    static getCallTraceByTransaction(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/{chainId}/transaction/{txHash}/call_trace',
            ...options
        });
    }
}
exports.DebugAndSimulationService = DebugAndSimulationService;
class ForksService {
    /**
     * List all forks
     */
    static listForks(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork',
            ...options
        });
    }
    /**
     * Create a fork
     */
    static createFork(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get trace by bundle simulation
     */
    static getCallTraceOnForkBundle(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{forkId}/bundle/{bundleId}/call_trace',
            ...options
        });
    }
    /**
     * Run Simulation
     */
    static simulateTransactionOnFork(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{forkId}/simulation',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get trace by simulation
     */
    static getCallTraceOnForkSimulation(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{forkId}/simulation/{simulationId}/call_trace',
            ...options
        });
    }
    /**
     * Run bundle simulation
     */
    static simulateTransactionBundleOnFork(options) {
        return (options.client ?? client_gen_1.client).post({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{forkId}/simulation_bundle',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get trace by transaction
     */
    static getCallTraceOnForkTransaction(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{forkId}/transaction/{txHash}/call_trace',
            ...options
        });
    }
    /**
     * Delete fork by id
     */
    static deleteFork(options) {
        return (options.client ?? client_gen_1.client).delete({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{id}',
            ...options
        });
    }
    /**
     * Get fork by id
     */
    static getFork(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{id}',
            ...options
        });
    }
    /**
     * Update fork by id
     */
    static updateFork(options) {
        return (options.client ?? client_gen_1.client).put({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{id}',
            ...options,
            headers: {
                'Content-Type': 'application/json',
                ...options?.headers
            }
        });
    }
    /**
     * Get fork info by id
     */
    static getForkInfo(options) {
        return (options.client ?? client_gen_1.client).get({
            security: [
                {
                    name: 'api-key',
                    type: 'apiKey'
                },
                {
                    in: 'query',
                    name: 'api-key',
                    type: 'apiKey'
                }
            ],
            url: '/api/v1/solidity/{owner}/{slug}/fork/{id}/info',
            ...options
        });
    }
}
exports.ForksService = ForksService;
