// @sentio/sdk/lib/eth/builtin/internal/eacaggregatorproxy-processor.js
import { addContractByABI, getContractByABI, addProcessor, getProcessor, getProvider, transformEtherError, BaseProcessor, BaseProcessorTemplate, BoundContractView, ContractView, DummyProvider, encodeCallData, } from "@sentio/sdk/eth";
import { EACAggregatorProxy__factory } from "./index.js";
const templateContract = EACAggregatorProxy__factory.connect("0x0", DummyProvider);
export class EACAggregatorProxyContractView extends ContractView {
    constructor(contract) {
        super(contract);
        this.callStatic.contract = contract;
    }
    async accessController(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("accessController()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async aggregator(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("aggregator()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async decimals(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("decimals()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async description(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("description()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async getAnswer(_roundId, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("getAnswer(uint256)")(_roundId, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async getRoundData(_roundId, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("getRoundData(uint80)")(_roundId, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async getTimestamp(_roundId, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("getTimestamp(uint256)")(_roundId, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async latestAnswer(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("latestAnswer()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async latestRound(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("latestRound()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async latestRoundData(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("latestRoundData()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async latestTimestamp(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("latestTimestamp()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async owner(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("owner()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async phaseAggregators(arg0, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("phaseAggregators(uint16)")(arg0, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async phaseId(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("phaseId()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async proposedAggregator(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("proposedAggregator()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async proposedGetRoundData(_roundId, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("proposedGetRoundData(uint80)")(_roundId, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async proposedLatestRoundData(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("proposedLatestRoundData()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async version(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("version()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    callStatic = {
        contract: this.contract,
        async acceptOwnership(overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("acceptOwnership()")
                    .staticCall(overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async confirmAggregator(_aggregator, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("confirmAggregator(address)")
                    .staticCall(_aggregator, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async proposeAggregator(_aggregator, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("proposeAggregator(address)")
                    .staticCall(_aggregator, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async setController(_accessController, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("setController(address)")
                    .staticCall(_accessController, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async transferOwnership(_to, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("transferOwnership(address)")
                    .staticCall(_to, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
    };
    encodeCall = {
        acceptOwnership(callContext) {
            return encodeCallData(callContext, "acceptOwnership", "function acceptOwnership()", []);
        },
        accessController(callContext) {
            return encodeCallData(callContext, "accessController", "function accessController()", []);
        },
        aggregator(callContext) {
            return encodeCallData(callContext, "aggregator", "function aggregator()", []);
        },
        confirmAggregator(_aggregator, callContext) {
            return encodeCallData(callContext, "confirmAggregator", "function confirmAggregator(address)", [_aggregator]);
        },
        decimals(callContext) {
            return encodeCallData(callContext, "decimals", "function decimals()", []);
        },
        description(callContext) {
            return encodeCallData(callContext, "description", "function description()", []);
        },
        getAnswer(_roundId, callContext) {
            return encodeCallData(callContext, "getAnswer", "function getAnswer(uint256)", [_roundId]);
        },
        getRoundData(_roundId, callContext) {
            return encodeCallData(callContext, "getRoundData", "function getRoundData(uint80)", [_roundId]);
        },
        getTimestamp(_roundId, callContext) {
            return encodeCallData(callContext, "getTimestamp", "function getTimestamp(uint256)", [_roundId]);
        },
        latestAnswer(callContext) {
            return encodeCallData(callContext, "latestAnswer", "function latestAnswer()", []);
        },
        latestRound(callContext) {
            return encodeCallData(callContext, "latestRound", "function latestRound()", []);
        },
        latestRoundData(callContext) {
            return encodeCallData(callContext, "latestRoundData", "function latestRoundData()", []);
        },
        latestTimestamp(callContext) {
            return encodeCallData(callContext, "latestTimestamp", "function latestTimestamp()", []);
        },
        owner(callContext) {
            return encodeCallData(callContext, "owner", "function owner()", []);
        },
        phaseAggregators(arg0, callContext) {
            return encodeCallData(callContext, "phaseAggregators", "function phaseAggregators(uint16)", [arg0]);
        },
        phaseId(callContext) {
            return encodeCallData(callContext, "phaseId", "function phaseId()", []);
        },
        proposeAggregator(_aggregator, callContext) {
            return encodeCallData(callContext, "proposeAggregator", "function proposeAggregator(address)", [_aggregator]);
        },
        proposedAggregator(callContext) {
            return encodeCallData(callContext, "proposedAggregator", "function proposedAggregator()", []);
        },
        proposedGetRoundData(_roundId, callContext) {
            return encodeCallData(callContext, "proposedGetRoundData", "function proposedGetRoundData(uint80)", [_roundId]);
        },
        proposedLatestRoundData(callContext) {
            return encodeCallData(callContext, "proposedLatestRoundData", "function proposedLatestRoundData()", []);
        },
        setController(_accessController, callContext) {
            return encodeCallData(callContext, "setController", "function setController(address)", [_accessController]);
        },
        transferOwnership(_to, callContext) {
            return encodeCallData(callContext, "transferOwnership", "function transferOwnership(address)", [_to]);
        },
        version(callContext) {
            return encodeCallData(callContext, "version", "function version()", []);
        },
    };
}
export class EACAggregatorProxyBoundContractView extends BoundContractView {
    async accessController(overrides) {
        return await this.view.accessController({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async aggregator(overrides) {
        return await this.view.aggregator({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async decimals(overrides) {
        return await this.view.decimals({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async description(overrides) {
        return await this.view.description({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async getAnswer(_roundId, overrides) {
        return await this.view.getAnswer(_roundId, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async getRoundData(_roundId, overrides) {
        return await this.view.getRoundData(_roundId, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async getTimestamp(_roundId, overrides) {
        return await this.view.getTimestamp(_roundId, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async latestAnswer(overrides) {
        return await this.view.latestAnswer({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async latestRound(overrides) {
        return await this.view.latestRound({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async latestRoundData(overrides) {
        return await this.view.latestRoundData({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async latestTimestamp(overrides) {
        return await this.view.latestTimestamp({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async owner(overrides) {
        return await this.view.owner({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async phaseAggregators(arg0, overrides) {
        return await this.view.phaseAggregators(arg0, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async phaseId(overrides) {
        return await this.view.phaseId({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async proposedAggregator(overrides) {
        return await this.view.proposedAggregator({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async proposedGetRoundData(_roundId, overrides) {
        return await this.view.proposedGetRoundData(_roundId, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async proposedLatestRoundData(overrides) {
        return await this.view.proposedLatestRoundData({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async version(overrides) {
        return await this.view.version({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    callStatic = {
        view: this.view,
        context: this.context,
        async acceptOwnership(overrides) {
            return await this.view.callStatic.acceptOwnership({
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async confirmAggregator(_aggregator, overrides) {
            return await this.view.callStatic.confirmAggregator(_aggregator, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async proposeAggregator(_aggregator, overrides) {
            return await this.view.callStatic.proposeAggregator(_aggregator, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async setController(_accessController, overrides) {
            return await this.view.callStatic.setController(_accessController, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async transferOwnership(_to, overrides) {
            return await this.view.callStatic.transferOwnership(_to, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
    };
    encodeCall = {
        view: this.view,
        context: this.context,
        acceptOwnership(overrides) {
            return this.view.encodeCall.acceptOwnership({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        accessController(overrides) {
            return this.view.encodeCall.accessController({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        aggregator(overrides) {
            return this.view.encodeCall.aggregator({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        confirmAggregator(_aggregator, overrides) {
            return this.view.encodeCall.confirmAggregator(_aggregator, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        decimals(overrides) {
            return this.view.encodeCall.decimals({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        description(overrides) {
            return this.view.encodeCall.description({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        getAnswer(_roundId, overrides) {
            return this.view.encodeCall.getAnswer(_roundId, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        getRoundData(_roundId, overrides) {
            return this.view.encodeCall.getRoundData(_roundId, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        getTimestamp(_roundId, overrides) {
            return this.view.encodeCall.getTimestamp(_roundId, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        latestAnswer(overrides) {
            return this.view.encodeCall.latestAnswer({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        latestRound(overrides) {
            return this.view.encodeCall.latestRound({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        latestRoundData(overrides) {
            return this.view.encodeCall.latestRoundData({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        latestTimestamp(overrides) {
            return this.view.encodeCall.latestTimestamp({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        owner(overrides) {
            return this.view.encodeCall.owner({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        phaseAggregators(arg0, overrides) {
            return this.view.encodeCall.phaseAggregators(arg0, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        phaseId(overrides) {
            return this.view.encodeCall.phaseId({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        proposeAggregator(_aggregator, overrides) {
            return this.view.encodeCall.proposeAggregator(_aggregator, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        proposedAggregator(overrides) {
            return this.view.encodeCall.proposedAggregator({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        proposedGetRoundData(_roundId, overrides) {
            return this.view.encodeCall.proposedGetRoundData(_roundId, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        proposedLatestRoundData(overrides) {
            return this.view.encodeCall.proposedLatestRoundData({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        setController(_accessController, overrides) {
            return this.view.encodeCall.setController(_accessController, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        transferOwnership(_to, overrides) {
            return this.view.encodeCall.transferOwnership(_to, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        version(overrides) {
            return this.view.encodeCall.version({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
    };
}
export class EACAggregatorProxyProcessor extends BaseProcessor {
    onEventAnswerUpdated(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["AnswerUpdated(int256,uint256,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventNewRound(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["NewRound(uint256,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferRequested(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferRequested(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferred(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferred(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onCallAcceptOwnership(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x79ba5097", handler, handlerOptions, preprocessHandler);
    }
    onCallAccessController(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xbc43cbaf", handler, handlerOptions, preprocessHandler);
    }
    onCallAggregator(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x245a7bfc", handler, handlerOptions, preprocessHandler);
    }
    onCallConfirmAggregator(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xa928c096", handler, handlerOptions, preprocessHandler);
    }
    onCallDecimals(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x313ce567", handler, handlerOptions, preprocessHandler);
    }
    onCallDescription(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x7284e416", handler, handlerOptions, preprocessHandler);
    }
    onCallGetAnswer(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xb5ab58dc", handler, handlerOptions, preprocessHandler);
    }
    onCallGetRoundData(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x9a6fc8f5", handler, handlerOptions, preprocessHandler);
    }
    onCallGetTimestamp(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xb633620c", handler, handlerOptions, preprocessHandler);
    }
    onCallLatestAnswer(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x50d25bcd", handler, handlerOptions, preprocessHandler);
    }
    onCallLatestRound(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x668a0f02", handler, handlerOptions, preprocessHandler);
    }
    onCallLatestRoundData(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xfeaf968c", handler, handlerOptions, preprocessHandler);
    }
    onCallLatestTimestamp(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x8205bf6a", handler, handlerOptions, preprocessHandler);
    }
    onCallOwner(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x8da5cb5b", handler, handlerOptions, preprocessHandler);
    }
    onCallPhaseAggregators(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xc1597304", handler, handlerOptions, preprocessHandler);
    }
    onCallPhaseId(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x58303b10", handler, handlerOptions, preprocessHandler);
    }
    onCallProposeAggregator(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xf8a2abd3", handler, handlerOptions, preprocessHandler);
    }
    onCallProposedAggregator(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xe8c4be30", handler, handlerOptions, preprocessHandler);
    }
    onCallProposedGetRoundData(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x6001ac53", handler, handlerOptions, preprocessHandler);
    }
    onCallProposedLatestRoundData(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x8f6b4d91", handler, handlerOptions, preprocessHandler);
    }
    onCallSetController(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x92eefe9b", handler, handlerOptions, preprocessHandler);
    }
    onCallTransferOwnership(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xf2fde38b", handler, handlerOptions, preprocessHandler);
    }
    onCallVersion(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x54fd4d50", handler, handlerOptions, preprocessHandler);
    }
    static filters = {
        AnswerUpdated(current, roundId, updatedAt) {
            return templateContract.filters["AnswerUpdated(int256,uint256,uint256)"](current, roundId, updatedAt);
        },
        NewRound(roundId, startedBy, startedAt) {
            return templateContract.filters["NewRound(uint256,address,uint256)"](roundId, startedBy, startedAt);
        },
        OwnershipTransferRequested(from, to) {
            return templateContract.filters["OwnershipTransferRequested(address,address)"](from, to);
        },
        OwnershipTransferred(from, to) {
            return templateContract.filters["OwnershipTransferred(address,address)"](from, to);
        },
    };
    CreateBoundContractView() {
        const view = getEACAggregatorProxyContract(this.config.network, this.config.address);
        return new EACAggregatorProxyBoundContractView(this.config.address, view);
    }
    static bind(options) {
        if (!options.name) {
            options.name = "EACAggregatorProxy";
        }
        let processor = getProcessor(options);
        if (!processor) {
            processor = new EACAggregatorProxyProcessor(options);
            addProcessor(options, processor);
        }
        return processor;
    }
}
export class EACAggregatorProxyProcessorTemplate extends BaseProcessorTemplate {
    bindInternal(options) {
        if (!options.name) {
            options.name = "EACAggregatorProxy";
        }
        let processor = getProcessor(options);
        if (!processor) {
            processor = new EACAggregatorProxyProcessor(options);
            addProcessor(options, processor);
        }
        return processor;
    }
    onEventAnswerUpdated(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["AnswerUpdated(int256,uint256,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventNewRound(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["NewRound(uint256,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferRequested(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferRequested(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferred(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferred(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
}
export function getEACAggregatorProxyContract(chainId, address) {
    let contract = getContractByABI("EACAggregatorProxy", address, chainId);
    if (!contract) {
        const rawContract = EACAggregatorProxy__factory.connect(address, getProvider(chainId));
        contract = new EACAggregatorProxyContractView(rawContract);
        addContractByABI("EACAggregatorProxy", address, chainId, contract);
    }
    return contract;
}
export function getEACAggregatorProxyContractOnContext(context, address) {
    const view = getEACAggregatorProxyContract(context.getChainId(), address);
    const boundView = new EACAggregatorProxyBoundContractView(address, view);
    boundView.context = context;
    if (boundView.callStatic) {
        boundView.callStatic.context = context;
    }
    return boundView;
}
