// @sentio/sdk/lib/eth/builtin/internal/erc20-processor.js
import { addContractByABI, getContractByABI, addProcessor, getProcessor, getProvider, transformEtherError, BaseProcessor, BaseProcessorTemplate, BoundContractView, ContractView, DummyProvider, encodeCallData, } from "@sentio/sdk/eth";
import { ERC20__factory } from "./index.js";
const templateContract = ERC20__factory.connect("0x0", DummyProvider);
export class ERC20ContractView extends ContractView {
    constructor(contract) {
        super(contract);
        this.callStatic.contract = contract;
    }
    async allowance(owner, spender, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("allowance(address,address)")(owner, spender, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async balanceOf(account, overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("balanceOf(address)")(account, overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async decimals(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("decimals()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async locker(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("locker()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async name(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("name()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async owner(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("owner()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async symbol(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("symbol()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    async totalSupply(overrides, preparedData, ethCallContext) {
        try {
            return await this.contract.getFunction("totalSupply()")(overrides || {});
        }
        catch (e) {
            const stack = new Error().stack;
            throw transformEtherError(e, undefined, stack);
        }
    }
    callStatic = {
        contract: this.contract,
        async approve(spender, amount, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("approve(address,uint256)")
                    .staticCall(spender, amount, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async burn(amount, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("burn(uint256)")
                    .staticCall(amount, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async burnFrom(account, amount, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("burnFrom(address,uint256)")
                    .staticCall(account, amount, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async decreaseAllowance(spender, subtractedValue, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("decreaseAllowance(address,uint256)")
                    .staticCall(spender, subtractedValue, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async increaseAllowance(spender, addedValue, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("increaseAllowance(address,uint256)")
                    .staticCall(spender, addedValue, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async renounceOwnership(overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("renounceOwnership()")
                    .staticCall(overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async setLocker(_locker, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("setLocker(address)")
                    .staticCall(_locker, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async transfer(recipient, amount, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("transfer(address,uint256)")
                    .staticCall(recipient, amount, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async transferFrom(sender, recipient, amount, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("transferFrom(address,address,uint256)")
                    .staticCall(sender, recipient, amount, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
        async transferOwnership(newOwner, overrides, preparedData, ethCallContext) {
            try {
                return await this.contract
                    .getFunction("transferOwnership(address)")
                    .staticCall(newOwner, overrides || {});
            }
            catch (e) {
                const stack = new Error().stack;
                throw transformEtherError(e, undefined, stack);
            }
        },
    };
    encodeCall = {
        allowance(owner, spender, callContext) {
            return encodeCallData(callContext, "allowance", "function allowance(address,address)", [owner, spender]);
        },
        approve(spender, amount, callContext) {
            return encodeCallData(callContext, "approve", "function approve(address,uint256)", [spender, amount]);
        },
        balanceOf(account, callContext) {
            return encodeCallData(callContext, "balanceOf", "function balanceOf(address)", [account]);
        },
        burn(amount, callContext) {
            return encodeCallData(callContext, "burn", "function burn(uint256)", [
                amount,
            ]);
        },
        burnFrom(account, amount, callContext) {
            return encodeCallData(callContext, "burnFrom", "function burnFrom(address,uint256)", [account, amount]);
        },
        decimals(callContext) {
            return encodeCallData(callContext, "decimals", "function decimals()", []);
        },
        decreaseAllowance(spender, subtractedValue, callContext) {
            return encodeCallData(callContext, "decreaseAllowance", "function decreaseAllowance(address,uint256)", [spender, subtractedValue]);
        },
        increaseAllowance(spender, addedValue, callContext) {
            return encodeCallData(callContext, "increaseAllowance", "function increaseAllowance(address,uint256)", [spender, addedValue]);
        },
        locker(callContext) {
            return encodeCallData(callContext, "locker", "function locker()", []);
        },
        name(callContext) {
            return encodeCallData(callContext, "name", "function name()", []);
        },
        owner(callContext) {
            return encodeCallData(callContext, "owner", "function owner()", []);
        },
        renounceOwnership(callContext) {
            return encodeCallData(callContext, "renounceOwnership", "function renounceOwnership()", []);
        },
        setLocker(_locker, callContext) {
            return encodeCallData(callContext, "setLocker", "function setLocker(address)", [_locker]);
        },
        symbol(callContext) {
            return encodeCallData(callContext, "symbol", "function symbol()", []);
        },
        totalSupply(callContext) {
            return encodeCallData(callContext, "totalSupply", "function totalSupply()", []);
        },
        transfer(recipient, amount, callContext) {
            return encodeCallData(callContext, "transfer", "function transfer(address,uint256)", [recipient, amount]);
        },
        transferFrom(sender, recipient, amount, callContext) {
            return encodeCallData(callContext, "transferFrom", "function transferFrom(address,address,uint256)", [sender, recipient, amount]);
        },
        transferOwnership(newOwner, callContext) {
            return encodeCallData(callContext, "transferOwnership", "function transferOwnership(address)", [newOwner]);
        },
    };
}
export class ERC20BoundContractView extends BoundContractView {
    async allowance(owner, spender, overrides) {
        return await this.view.allowance(owner, spender, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async balanceOf(account, overrides) {
        return await this.view.balanceOf(account, {
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async decimals(overrides) {
        return await this.view.decimals({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async locker(overrides) {
        return await this.view.locker({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async name(overrides) {
        return await this.view.name({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async owner(overrides) {
        return await this.view.owner({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async symbol(overrides) {
        return await this.view.symbol({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    async totalSupply(overrides) {
        return await this.view.totalSupply({
            blockTag: this.context.blockNumber,
            ...overrides,
        }, this.context.preparedData, this.context.getEthCallContext());
    }
    callStatic = {
        view: this.view,
        context: this.context,
        async approve(spender, amount, overrides) {
            return await this.view.callStatic.approve(spender, amount, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async burn(amount, overrides) {
            return await this.view.callStatic.burn(amount, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async burnFrom(account, amount, overrides) {
            return await this.view.callStatic.burnFrom(account, amount, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async decreaseAllowance(spender, subtractedValue, overrides) {
            return await this.view.callStatic.decreaseAllowance(spender, subtractedValue, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async increaseAllowance(spender, addedValue, overrides) {
            return await this.view.callStatic.increaseAllowance(spender, addedValue, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async renounceOwnership(overrides) {
            return await this.view.callStatic.renounceOwnership({
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async setLocker(_locker, overrides) {
            return await this.view.callStatic.setLocker(_locker, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async transfer(recipient, amount, overrides) {
            return await this.view.callStatic.transfer(recipient, amount, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async transferFrom(sender, recipient, amount, overrides) {
            return await this.view.callStatic.transferFrom(sender, recipient, amount, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
        async transferOwnership(newOwner, overrides) {
            return await this.view.callStatic.transferOwnership(newOwner, {
                blockTag: this.context.blockNumber,
                ...overrides,
            }, this.context.preparedData, this.context.getEthCallContext());
        },
    };
    encodeCall = {
        view: this.view,
        context: this.context,
        allowance(owner, spender, overrides) {
            return this.view.encodeCall.allowance(owner, spender, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        approve(spender, amount, overrides) {
            return this.view.encodeCall.approve(spender, amount, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        balanceOf(account, overrides) {
            return this.view.encodeCall.balanceOf(account, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        burn(amount, overrides) {
            return this.view.encodeCall.burn(amount, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        burnFrom(account, amount, overrides) {
            return this.view.encodeCall.burnFrom(account, amount, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        decimals(overrides) {
            return this.view.encodeCall.decimals({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        decreaseAllowance(spender, subtractedValue, overrides) {
            return this.view.encodeCall.decreaseAllowance(spender, subtractedValue, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        increaseAllowance(spender, addedValue, overrides) {
            return this.view.encodeCall.increaseAllowance(spender, addedValue, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        locker(overrides) {
            return this.view.encodeCall.locker({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        name(overrides) {
            return this.view.encodeCall.name({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        owner(overrides) {
            return this.view.encodeCall.owner({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        renounceOwnership(overrides) {
            return this.view.encodeCall.renounceOwnership({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        setLocker(_locker, overrides) {
            return this.view.encodeCall.setLocker(_locker, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        symbol(overrides) {
            return this.view.encodeCall.symbol({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        totalSupply(overrides) {
            return this.view.encodeCall.totalSupply({
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        transfer(recipient, amount, overrides) {
            return this.view.encodeCall.transfer(recipient, amount, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        transferFrom(sender, recipient, amount, overrides) {
            return this.view.encodeCall.transferFrom(sender, recipient, amount, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
        transferOwnership(newOwner, overrides) {
            return this.view.encodeCall.transferOwnership(newOwner, {
                chainId: this.context.chainId.toString(),
                address: this.context.address,
                blockTag: this.context.getBlockTag(overrides),
            });
        },
    };
}
export class ERC20Processor extends BaseProcessor {
    onEventApproval(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["Approval(address,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferred(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferred(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventTransfer(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["Transfer(address,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onCallAllowance(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xdd62ed3e", handler, handlerOptions, preprocessHandler);
    }
    onCallApprove(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x095ea7b3", handler, handlerOptions, preprocessHandler);
    }
    onCallBalanceOf(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x70a08231", handler, handlerOptions, preprocessHandler);
    }
    onCallBurn(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x42966c68", handler, handlerOptions, preprocessHandler);
    }
    onCallBurnFrom(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x79cc6790", handler, handlerOptions, preprocessHandler);
    }
    onCallDecimals(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x313ce567", handler, handlerOptions, preprocessHandler);
    }
    onCallDecreaseAllowance(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xa457c2d7", handler, handlerOptions, preprocessHandler);
    }
    onCallIncreaseAllowance(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x39509351", handler, handlerOptions, preprocessHandler);
    }
    onCallLocker(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xd7b96d4e", handler, handlerOptions, preprocessHandler);
    }
    onCallName(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x06fdde03", handler, handlerOptions, preprocessHandler);
    }
    onCallOwner(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x8da5cb5b", handler, handlerOptions, preprocessHandler);
    }
    onCallRenounceOwnership(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x715018a6", handler, handlerOptions, preprocessHandler);
    }
    onCallSetLocker(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x171060ec", handler, handlerOptions, preprocessHandler);
    }
    onCallSymbol(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x95d89b41", handler, handlerOptions, preprocessHandler);
    }
    onCallTotalSupply(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x18160ddd", handler, handlerOptions, preprocessHandler);
    }
    onCallTransfer(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xa9059cbb", handler, handlerOptions, preprocessHandler);
    }
    onCallTransferFrom(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0x23b872dd", handler, handlerOptions, preprocessHandler);
    }
    onCallTransferOwnership(handler, handlerOptions, preprocessHandler) {
        return super.onEthTrace("0xf2fde38b", handler, handlerOptions, preprocessHandler);
    }
    static filters = {
        Approval(owner, spender, value) {
            return templateContract.filters["Approval(address,address,uint256)"](owner, spender, value);
        },
        OwnershipTransferred(previousOwner, newOwner) {
            return templateContract.filters["OwnershipTransferred(address,address)"](previousOwner, newOwner);
        },
        Transfer(from, to, value) {
            return templateContract.filters["Transfer(address,address,uint256)"](from, to, value);
        },
    };
    CreateBoundContractView() {
        const view = getERC20Contract(this.config.network, this.config.address);
        return new ERC20BoundContractView(this.config.address, view);
    }
    static bind(options) {
        if (!options.name) {
            options.name = "ERC20";
        }
        let processor = getProcessor(options);
        if (!processor) {
            processor = new ERC20Processor(options);
            addProcessor(options, processor);
        }
        return processor;
    }
}
export class ERC20ProcessorTemplate extends BaseProcessorTemplate {
    bindInternal(options) {
        if (!options.name) {
            options.name = "ERC20";
        }
        let processor = getProcessor(options);
        if (!processor) {
            processor = new ERC20Processor(options);
            addProcessor(options, processor);
        }
        return processor;
    }
    onEventApproval(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["Approval(address,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventOwnershipTransferred(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["OwnershipTransferred(address,address)"](null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
    onEventTransfer(handler, filter, handlerOptions, preprocessHandler) {
        if (!filter) {
            filter = templateContract.filters["Transfer(address,address,uint256)"](null, null, null);
        }
        return super.onEthEvent(handler, filter, handlerOptions, preprocessHandler);
    }
}
export function getERC20Contract(chainId, address) {
    let contract = getContractByABI("ERC20", address, chainId);
    if (!contract) {
        const rawContract = ERC20__factory.connect(address, getProvider(chainId));
        contract = new ERC20ContractView(rawContract);
        addContractByABI("ERC20", address, chainId, contract);
    }
    return contract;
}
export function getERC20ContractOnContext(context, address) {
    const view = getERC20Contract(context.getChainId(), address);
    const boundView = new ERC20BoundContractView(address, view);
    boundView.context = context;
    if (boundView.callStatic) {
        boundView.callStatic.context = context;
    }
    return boundView;
}
