// @typemove/iota/dist/cjs/transaction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTransactionArgument = isTransactionArgument;
exports.transactionArgumentOrObject = transactionArgumentOrObject;
exports.transactionArgumentOrPure = transactionArgumentOrPure;
exports.transactionArgumentOrPureString = transactionArgumentOrPureString;
exports.transactionArgumentOrPureAddress = transactionArgumentOrPureAddress;
exports.transactionArgumentOrPureU8 = transactionArgumentOrPureU8;
exports.transactionArgumentOrPureU16 = transactionArgumentOrPureU16;
exports.transactionArgumentOrPureU32 = transactionArgumentOrPureU32;
exports.transactionArgumentOrPureU64 = transactionArgumentOrPureU64;
exports.transactionArgumentOrPureU128 = transactionArgumentOrPureU128;
exports.transactionArgumentOrPureU256 = transactionArgumentOrPureU256;
exports.transactionArgumentOrPureBool = transactionArgumentOrPureBool;
exports.transactionArgumentOrVec = transactionArgumentOrVec;
function isTransactionArgument(value) {
    if (typeof value !== 'object')
        return false;
    if (value === null || value === undefined)
        return false;
    return (value.$kind === 'GasCoin' || value.$kind === 'Result' || value.$kind === 'NestedResult' || value.$kind === 'Input');
}
function transactionArgumentOrObject(value, transactionBlock) {
    // if (isTransactionArgument(value)) {
    //   return value as TransactionArgument
    // }
    return transactionBlock.object(value);
}
function transactionArgumentOrPure(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return typeof value == 'string' ? transactionBlock.pure.string(value) : transactionBlock.pure.u64(value);
}
function transactionArgumentOrPureString(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.string(value);
}
function transactionArgumentOrPureAddress(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.address(value);
}
function transactionArgumentOrPureU8(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u8(value);
}
function transactionArgumentOrPureU16(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u16(value);
}
function transactionArgumentOrPureU32(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u32(value);
}
function transactionArgumentOrPureU64(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u64(value);
}
function transactionArgumentOrPureU128(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u128(value);
}
function transactionArgumentOrPureU256(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u256(value);
}
function transactionArgumentOrPureBool(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.bool(value);
}
// TODO vector might be nested
function transactionArgumentOrVec(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.makeMoveVec({
        elements: value.map((a) => transactionBlock.object(a))
    });
}
