// @sentio/sdk/lib/eth/abi-decoder/decode-worker.js
import { Interface } from 'ethers';
import { Piscina } from 'piscina';
import { ServerError, Status } from 'nice-grpc';
import { FormattedLog } from '../eth.js';
export function decodeTraceInline(contractViewInterface, inputs, traceData) {
    return contractViewInterface.getAbiCoder().decode(inputs, traceData, true);
}
export function decodeTrace({ inputs, traceData }) {
    const fragments = Piscina['workerData'];
    const contractViewInterface = new Interface(fragments);
    const result = decodeTraceInline(contractViewInterface, inputs, traceData);
    const argsObj = result?.args.toObject(true);
    const argsArray = [];
    const keys = [];
    for (const [key, value] of Object.entries(argsObj)) {
        keys.push(key);
        argsArray.push(value);
    }
    return {
        array: argsArray,
        keys: keys
    };
}
export function parseLogInline(contractViewInterface, log) {
    if (!log) {
        throw new ServerError(Status.INVALID_ARGUMENT, 'Log is empty');
    }
    const topics = log.topics;
    const data = log.data;
    return contractViewInterface.parseLog({ topics, data });
}
export function parseLog({ log }) {
    const fragments = Piscina['workerData'];
    const contractViewInterface = new Interface(fragments);
    if (log.raw) {
        log = new FormattedLog(log.raw);
    }
    const result = parseLogInline(contractViewInterface, log);
    if (!result) {
        return null;
    }
    // can't serialize LogDescription, so return args and topics only
    if (result?.args == null) {
        return {
            args: null,
            topic: result.topic
        };
    }
    const argsObj = result?.args.toObject(true);
    const argsArray = [];
    const keys = [];
    for (const [key, value] of Object.entries(argsObj)) {
        keys.push(key);
        argsArray.push(value);
    }
    return {
        args: {
            array: argsArray,
            keys: keys
        },
        topic: result.topic
    };
}
export default {
    parseLog,
    decodeTrace
};
