// @sentio/chain/dist/chain-id.js
"use strict";
// copy from https://github.com/DefiLlama/chainlist/blob/main/constants/chainIds.js
// and https://besu.hyperledger.org/en/stable/Concepts/NetworkID-And-ChainID/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainIdToType = exports.ChainTypeToChainId = exports.ChainId = exports.NonEthChainId = exports.BTCChainId = exports.StarknetChainId = exports.CosmosChainId = exports.FuelChainId = exports.SolanaChainId = exports.SuiChainId = exports.AptosChainId = exports.EthChainId = exports.ChainType = void 0;
exports.getChainType = getChainType;
exports.isChainType = isChainType;
var ChainType;
(function (ChainType) {
    ChainType["SOLANA"] = "solana";
    ChainType["SUI"] = "sui";
    ChainType["APTOS"] = "aptos";
    ChainType["ETH"] = "evm";
    ChainType["BTC"] = "btc";
    ChainType["COSMOS"] = "cosmos";
    ChainType["STARKNET"] = "starknet";
    ChainType["FUEL"] = "fuel";
})(ChainType || (exports.ChainType = ChainType = {}));
var EthChainId;
(function (EthChainId) {
    // Any modify to this must also be modify in EthChainName
    EthChainId["ETHEREUM"] = "1";
    EthChainId["OPTIMISM"] = "10";
    EthChainId["CRONOS"] = "25";
    EthChainId["BSC"] = "56";
    EthChainId["BSC_TESTNET"] = "97";
    EthChainId["UNICHAIN"] = "130";
    EthChainId["POLYGON"] = "137";
    EthChainId["MANTA_PACIFIC"] = "169";
    EthChainId["XLAYER_TESTNET"] = "195";
    EthChainId["XLAYER_MAINNET"] = "196";
    EthChainId["OP_BNB_MAINNET"] = "204";
    EthChainId["SONIC_MAINNET"] = "146";
    EthChainId["SONIC_TESTNET"] = "14601";
    EthChainId["B2_MAINNET"] = "223";
    // FANTOM = '250',
    EthChainId["FRAXTAL"] = "252";
    EthChainId["KUCOIN"] = "321";
    EthChainId["CRONOS_ZKEVM"] = "388";
    EthChainId["ZKSYNC_ERA"] = "324";
    EthChainId["CRONOS_TESTNET"] = "338";
    EthChainId["ASTAR"] = "592";
    EthChainId["DERIVE"] = "957";
    EthChainId["STABLE_TESTNET"] = "2201";
    EthChainId["STABLE_MAINNET"] = "988";
    EthChainId["HYPER_EVM"] = "999";
    EthChainId["POLYGON_ZKEVM"] = "1101";
    EthChainId["CONFLUX"] = "1030";
    EthChainId["METIS"] = "1088";
    EthChainId["CORE_MAINNET"] = "1116";
    EthChainId["MOONBEAM"] = "1284";
    EthChainId["UNICHAIN_SEPOLIA"] = "1301";
    EthChainId["SEI"] = "1329";
    EthChainId["SONEIUM_TESTNET"] = "1946";
    EthChainId["SONEIUM_MAINNET"] = "1868";
    EthChainId["SWELL_TESTNET"] = "1924";
    EthChainId["SWELL_MAINNET"] = "1923";
    EthChainId["TAC_TESTNET"] = "2390";
    EthChainId["KARAK"] = "2410";
    EthChainId["BEVM"] = "11501";
    EthChainId["MERLIN_MAINNET"] = "4200";
    EthChainId["MANTLE"] = "5000";
    EthChainId["BASE"] = "8453";
    EthChainId["BASE_SEPOLIA"] = "84532";
    EthChainId["MEV_COMMIT"] = "57173";
    EthChainId["PLASMA_MAINNET"] = "9745";
    EthChainId["PLASMA_TESTNET"] = "9746";
    EthChainId["MONAD_TESTNET"] = "10143";
    EthChainId["MONAD_MAINNET"] = "143";
    EthChainId["HOLESKY"] = "17000";
    EthChainId["HOODI"] = "560048";
    EthChainId["MODE"] = "34443";
    EthChainId["ARBITRUM"] = "42161";
    EthChainId["HEMI"] = "43111";
    EthChainId["AVALANCHE"] = "43114";
    EthChainId["ZIRCUIT_GARFIELD_TESTNET"] = "48898";
    EthChainId["ZIRCUIT_MAINNET"] = "48900";
    EthChainId["BOB"] = "60808";
    EthChainId["LINEA"] = "59144";
    EthChainId["BERACHAIN"] = "80094";
    EthChainId["BLAST"] = "81457";
    EthChainId["CHILIZ"] = "88888";
    EthChainId["TAIKO"] = "167000";
    EthChainId["KATANA_MAINNET"] = "747474";
    // TAIKO_TESTNET = '167009',
    EthChainId["BITLAYER"] = "200901";
    EthChainId["SCROLL"] = "534352";
    EthChainId["ZKLINK_NOVA"] = "810180";
    EthChainId["SEPOLIA"] = "11155111";
    EthChainId["ETHERLINK"] = "42793";
    EthChainId["CORN_MAIZENET"] = "21000000";
    EthChainId["AURORA"] = "1313161554";
    EthChainId["BLAST_SEPOLIA"] = "168587773";
    EthChainId["ABSTRACT"] = "2741";
    EthChainId["GOAT_MAINNET"] = "2345";
    EthChainId["ARC_TESTNET"] = "5042002";
    EthChainId["SENTIO_MAINNET"] = "789210";
    EthChainId["SENTIO_TESTNET"] = "7892101";
    EthChainId["CUSTOM"] = "customized";
})(EthChainId || (exports.EthChainId = EthChainId = {}));
var AptosChainId;
(function (AptosChainId) {
    AptosChainId["APTOS_MAINNET"] = "aptos_mainnet";
    AptosChainId["APTOS_TESTNET"] = "aptos_testnet";
    AptosChainId["APTOS_MOVEMENT_TESTNET"] = "aptos_movement_testnet";
    AptosChainId["APTOS_MOVEMENT_MAINNET"] = "aptos_movement_mainnet";
    AptosChainId["APTOS_MOVEMENT_PREVIEWNET"] = "aptos_movement_previewnet";
    AptosChainId["INITIA_ECHELON"] = "aptos_echelon";
})(AptosChainId || (exports.AptosChainId = AptosChainId = {}));
var SuiChainId;
(function (SuiChainId) {
    SuiChainId["SUI_MAINNET"] = "sui_mainnet";
    SuiChainId["SUI_TESTNET"] = "sui_testnet";
    SuiChainId["IOTA_MAINNET"] = "iota_mainnet";
    SuiChainId["IOTA_TESTNET"] = "iota_testnet";
})(SuiChainId || (exports.SuiChainId = SuiChainId = {}));
var SolanaChainId;
(function (SolanaChainId) {
    SolanaChainId["SOLANA_MAINNET"] = "sol_mainnet";
    // SOLANA_DEVNET = 'sol_devnet',
    SolanaChainId["SOLANA_TESTNET"] = "sol_testnet";
    SolanaChainId["SOLANA_PYTH"] = "sol_pyth";
    SolanaChainId["FORGO_TESTNET"] = "forgo_testnet";
    SolanaChainId["FORGO_MAINNET"] = "forgo_mainnet";
})(SolanaChainId || (exports.SolanaChainId = SolanaChainId = {}));
var FuelChainId;
(function (FuelChainId) {
    FuelChainId["FUEL_MAINNET"] = "fuel_mainnet";
    FuelChainId["FUEL_TESTNET"] = "fuel_testnet";
})(FuelChainId || (exports.FuelChainId = FuelChainId = {}));
var CosmosChainId;
(function (CosmosChainId) {
    CosmosChainId["INJECTIVE_MAINNET"] = "injective_mainnet";
    CosmosChainId["INJECTIVE_TESTNET"] = "injective_testnet";
})(CosmosChainId || (exports.CosmosChainId = CosmosChainId = {}));
var StarknetChainId;
(function (StarknetChainId) {
    StarknetChainId["STARKNET_MAINNET"] = "starknet_mainnet";
    StarknetChainId["STARKNET_SEPOLIA"] = "starknet_sepolia";
})(StarknetChainId || (exports.StarknetChainId = StarknetChainId = {}));
var BTCChainId;
(function (BTCChainId) {
    BTCChainId["BTC_MAINNET"] = "btc_mainnet";
    BTCChainId["BTC_TESTNET"] = "btc_testnet";
})(BTCChainId || (exports.BTCChainId = BTCChainId = {}));
exports.NonEthChainId = {
    ...AptosChainId,
    ...SuiChainId,
    ...SolanaChainId,
    ...FuelChainId,
    ...CosmosChainId,
    ...StarknetChainId,
    ...BTCChainId
};
exports.ChainId = {
    ...EthChainId,
    ...exports.NonEthChainId
};
exports.ChainTypeToChainId = {
    [ChainType.SOLANA]: SolanaChainId,
    [ChainType.SUI]: SuiChainId,
    [ChainType.COSMOS]: CosmosChainId,
    [ChainType.STARKNET]: StarknetChainId,
    [ChainType.ETH]: EthChainId,
    [ChainType.APTOS]: AptosChainId,
    [ChainType.BTC]: BTCChainId,
    [ChainType.FUEL]: FuelChainId
};
exports.ChainIdToType = new Map();
for (const [chainType, chainId] of Object.entries(exports.ChainTypeToChainId)) {
    for (const value of Object.values(chainId)) {
        exports.ChainIdToType.set(value, chainType);
    }
}
function getChainType(chainId) {
    const id = String(chainId).toLowerCase();
    const chainType = exports.ChainIdToType.get(id);
    if (!chainType) {
        throw new Error(`Invalid chainType: ${id}`);
    }
    return chainType;
}
function isChainType(chainId, targetChainType) {
    const id = String(chainId).toLowerCase();
    const chainType = exports.ChainIdToType.get(id);
    if (!chainType) {
        return false;
    }
    return chainType === targetChainType;
}
