// @sentio/sdk/lib/aptos/api.js
import { Aptos } from '@aptos-labs/ts-sdk';
import { defaultMoveCoderForClient } from './move-coder.js';
// Sentinel value to indicate "latest" ledger version
export const LATEST_LEDGER_VERSION = Number.MAX_SAFE_INTEGER;
// Aptos Client inherit [[ Aptos ]] and add getTypedAccountResource method with could get resource and
// decode it to the specified type
export class RichAptosClient extends Aptos {
    async getTypedAccountResource(args) {
        const type = args.resourceType;
        const res = await this.getAccountResource({
            ...args,
            resourceType: type.getSignature()
        });
        return (await defaultMoveCoderForClient(this).decodeType(res, type));
    }
}
// Aptos Client inherit [[ RichAptosClient ]] and add context to it
// will inject ledger version from context for the following methods if ledgerVersion not specified
// - getAccountResource(s)
// - getModule(s)
// - view/viewJson
export class RichAptosClientWithContext extends RichAptosClient {
    ctx;
    constructor(ctx, config) {
        super(config);
        this.ctx = ctx;
    }
    transformArgs(args) {
        // Check if ledgerVersion was explicitly provided
        if (!args.options || !('ledgerVersion' in args.options)) {
            // No ledgerVersion specified at all - use context version
            args.options = {
                ...args.options,
                ledgerVersion: this.ctx.version
            };
        }
        else if (args.options.ledgerVersion === LATEST_LEDGER_VERSION) {
            // Explicitly requested latest state using sentinel value
            args.options = {
                ...args.options,
                ledgerVersion: undefined // This tells Aptos SDK to use latest
            };
        }
        // Otherwise, use the provided ledgerVersion as-is
        return args;
    }
    view(args) {
        return super.view(this.transformArgs(args));
    }
    viewJson(args) {
        return super.viewJson(this.transformArgs(args));
    }
    getAccountResource(args) {
        return super.getAccountResource(this.transformArgs(args));
    }
    getAccountResources(args) {
        return super.getAccountResources(this.transformArgs(args));
    }
    getAccountModules(args) {
        return super.getAccountModules(this.transformArgs(args));
    }
    getAccountModule(args) {
        return super.getAccountModule(this.transformArgs(args));
    }
    lookupOriginalAccountAddress(args) {
        return super.lookupOriginalAccountAddress(this.transformArgs(args));
    }
}
