// @sentio/sdk/lib/aptos/aptos-resource-processor-template.js
import { ListStateStorage, processMetrics } from '@sentio/runtime';
import { AptosResourcesProcessor, DEFAULT_RESOURCE_FETCH_CONFIG } from './aptos-processor.js';
import { getHandlerName, proxyProcessor } from '../utils/metrics.js';
export class AptosResourceProcessorTemplateState extends ListStateStorage {
    static INSTANCE = new AptosResourceProcessorTemplateState();
}
class Handler {
    type;
    checkpointInterval;
    timeIntervalInMinutes;
    handlerName;
    handler;
    handlerOptions;
}
export class AptosResourceProcessorTemplate {
    id;
    handlers = [];
    instances = new Set();
    constructor() {
        this.id = AptosResourceProcessorTemplateState.INSTANCE.getValues().length;
        AptosResourceProcessorTemplateState.INSTANCE.addValue(this);
        return proxyProcessor(this);
    }
    createProcessor(options) {
        return AptosResourcesProcessor.bind(options);
    }
    bind(options, ctx) {
        options.network = options.network || ctx.network;
        options.startVersion = options.startVersion || ctx.version;
        const instance = {
            templateId: this.id,
            contract: {
                name: '',
                chainId: options.network,
                address: options.address,
                abi: ''
            },
            startBlock: options.startVersion ? BigInt(options.startVersion) : 0n,
            endBlock: options.endVersion ? BigInt(options.endVersion) : 0n,
            baseLabels: options.baseLabels
        };
        ctx.sendTemplateInstance(instance);
        ctx.update({
            states: {
                configUpdated: true
            }
        });
        processMetrics.processor_template_instance_count.add(1, {
            chain_id: options.network,
            template: this.constructor.name
        });
    }
    startInstance(options, ctx) {
        options.network = options.network || ctx.network;
        options.startVersion = options.startVersion || ctx.version;
        const id = options.address;
        const sig = [options.network, id].join('_');
        if (this.instances.has(sig)) {
            console.debug(`Same object id can be bind to one template only once, ignore duplicate bind: ${sig}`);
            return;
        }
        this.instances.add(sig);
        const processor = this.createProcessor(options);
        for (const h of this.handlers) {
            processor.onInterval(h.handler, h.timeIntervalInMinutes, h.checkpointInterval, h.type, h.handlerOptions, h.handlerName);
        }
    }
    onInterval(handler, timeInterval, checkpointInterval, type, handlerOptions) {
        this.handlers.push({
            handlerName: getHandlerName(),
            handler: handler,
            timeIntervalInMinutes: timeInterval,
            checkpointInterval: checkpointInterval,
            type,
            handlerOptions: { ...DEFAULT_RESOURCE_FETCH_CONFIG, ...handlerOptions }
        });
        return this;
    }
    onTimeInterval(handler, timeIntervalInMinutes = 60, backfillTimeIntervalInMinutes = 240, type, handlerOptions) {
        return this.onInterval(handler, {
            recentInterval: timeIntervalInMinutes,
            backfillInterval: backfillTimeIntervalInMinutes
        }, undefined, type, handlerOptions);
    }
    onVersionInterval(handler, checkpointInterval = 100000, backfillCheckpointInterval = 400000, type, handlerOptions) {
        return this.onInterval(handler, undefined, { recentInterval: checkpointInterval, backfillInterval: backfillCheckpointInterval }, type, handlerOptions);
    }
}
