// @sentio/sdk/lib/aptos/context.js
import { normalizeLabels } from '../index.js';
import { RichAptosClientWithContext } from './index.js';
import { AptosConfig, Network } from '@aptos-labs/ts-sdk';
import { defaultMoveCoder } from './move-coder.js';
import { AptosNetwork } from './network.js';
import { Endpoints } from '@sentio/runtime';
import { ServerError, Status } from 'nice-grpc';
import { MoveContext } from '../move/index.js';
export class AptosBaseContext extends MoveContext {
    version;
    coder;
    constructor(network, version, labels) {
        super(labels);
        this.network = network;
        this.coder = defaultMoveCoder(this.network);
        this.version = version;
    }
    getClient() {
        const chainServer = Endpoints.INSTANCE.chainServer.get(this.network);
        if (!chainServer) {
            throw new ServerError(Status.INTERNAL, 'RPC endpoint not provided');
        }
        const fullnode = chainServer + '/v1';
        let network = Network.CUSTOM;
        switch (this.network) {
            case AptosNetwork.MAIN_NET:
                network = Network.MAINNET;
                break;
            case AptosNetwork.TEST_NET:
                network = Network.TESTNET;
                break;
        }
        return new RichAptosClientWithContext(this, new AptosConfig({
            network,
            fullnode
        }));
    }
}
export class AptosTransactionContext extends AptosBaseContext {
    moduleName;
    transaction;
    eventIndex;
    constructor(moduleName, network, address, version, transaction, eventIndex, baseLabels) {
        super(network, version, baseLabels);
        this.address = address.toLowerCase();
        this.moduleName = moduleName;
        this.eventIndex = eventIndex;
        if (transaction) {
            this.transaction = transaction;
        }
    }
    getChainId() {
        return this.network;
    }
    getTimestamp() {
        return parseInt(this.transaction.timestamp);
    }
    getMetaDataInternal(name, labels) {
        return {
            address: this.address,
            contractName: this.moduleName,
            blockNumber: this.version,
            transactionIndex: 0,
            transactionHash: this.transaction?.hash || '', // TODO
            logIndex: this.eventIndex,
            chainId: this.getChainId(),
            name: name,
            labels: normalizeLabels(labels)
        };
    }
}
export class AptosContext extends AptosTransactionContext {
}
export class AptosResourcesContext extends AptosBaseContext {
    timestampInMicros;
    constructor(network, address, version, timestampInMicros, baseLabels) {
        super(network, version, baseLabels);
        this.address = address;
        this.timestampInMicros = timestampInMicros;
    }
    getChainId() {
        return this.network;
    }
    getTimestamp() {
        return this.timestampInMicros;
    }
    getMetaDataInternal(name, labels) {
        return {
            address: this.address,
            contractName: 'resources',
            blockNumber: this.version,
            transactionIndex: 0,
            transactionHash: '',
            logIndex: 0,
            chainId: this.getChainId(),
            name: name,
            labels: normalizeLabels(labels)
        };
    }
}
