// @sentio/sdk/lib/aptos/data.js
import { decodeResourceChange } from '@typemove/aptos';
export class AptEvent {
    data;
    constructor(data) {
        this.data = data;
    }
    get rawEvent() {
        return this.data.rawEvent;
    }
    get eventIndex() {
        return this.data.eventIndex;
    }
    get rawTransaction() {
        return this.data.rawTransaction;
    }
    get transaction() {
        if (!this.data._tx) {
            this.data._tx = JSON.parse(this.data.rawTransaction);
            if (this.data._tx.events == null) {
                this.data._tx.events = [];
            }
        }
        return this.data._tx;
    }
    get event() {
        if (!this.data._event) {
            this.data._event = JSON.parse(this.data.rawEvent);
        }
        return this.data._event;
    }
    async decodeEvent(coder) {
        if (!this.data._decodedEvent) {
            this.data._decodedEvent = await coder.decodeEvent(this.event);
            return this.data._decodedEvent;
        }
        return this.data._decodedEvent;
    }
}
export class AptCall {
    data;
    constructor(data) {
        this.data = data;
    }
    get rawTransaction() {
        return this.data.rawTransaction;
    }
    get transaction() {
        if (!this.data._tx) {
            this.data._tx = JSON.parse(this.data.rawTransaction);
            if (this.data._tx.events == null) {
                this.data._tx.events = [];
            }
        }
        return this.data._tx;
    }
    async decodeCall(coder) {
        if (!this.data._decodedCall) {
            const tx = this.transaction;
            let payload = tx.payload;
            if (payload.type === 'multisig_payload') {
                payload = payload.transaction_payload ?? payload;
            }
            this.data._decodedCall = await coder.decodeFunctionPayload(payload);
            return this.data._decodedCall;
        }
        return this.data._decodedCall;
    }
}
export class AptResource {
    data;
    constructor(data) {
        this.data = data;
    }
    get version() {
        return this.data.version;
    }
    get timestampMicros() {
        return this.data.timestampMicros;
    }
    get rawResources() {
        return this.data.rawResources;
    }
    get resources() {
        if (!this.data._resources) {
            this.data._resources = this.data.rawResources.map((r) => JSON.parse(r));
        }
        return this.data._resources;
    }
    get timestamp() {
        return new Date(Number(this.data.timestampMicros) / 1000);
    }
    async decodeResources(coder) {
        if (!this.data._decodedResources) {
            this.data._decodedResources = await decodeResourceChange(this.resources, coder);
        }
        return this.data._decodedResources;
    }
}
