// @sentio/sdk/lib/fuel/fuel-processor-template.js
import { ListStateStorage, processMetrics } from '@sentio/runtime';
import { DEFAULT_FUEL_FETCH_CONFIG } from './transaction.js';
import { getOptionsSignature } from './fuel-processor.js';
import { getHandlerName, proxyProcessor } from '../utils/metrics.js';
export class FuelProcessorTemplateProcessorState extends ListStateStorage {
    static INSTANCE = new FuelProcessorTemplateProcessorState();
}
export class FuelBaseProcessorTemplate {
    id;
    instances = new Set();
    blockHandlers = [];
    logHandlers = [];
    transactionHandlers = [];
    constructor() {
        this.id = FuelProcessorTemplateProcessorState.INSTANCE.getValues().length;
        FuelProcessorTemplateProcessorState.INSTANCE.addValue(this);
        return proxyProcessor(this);
    }
    /**
     * Bind template using {@param options}, using {@param ctx}'s network value if not provided in the option
     * @param options
     * @param ctx
     */
    bind(options, ctx) {
        const instance = {
            templateId: this.id,
            contract: {
                address: options.address,
                name: options.name || '',
                chainId: ctx.chainId,
                abi: ''
            },
            startBlock: BigInt(options.startBlock || 0),
            endBlock: BigInt(options.endBlock || 0),
            baseLabels: {}
            // baseLabels: options.baseLabels
        };
        ctx.sendTemplateInstance(instance);
        ctx.update({
            states: {
                configUpdated: true
            }
        });
        processMetrics.processor_template_instance_count.add(1, {
            chain_id: ctx.chainId,
            template: this.constructor.name
        });
    }
    startInstance(options, ctx) {
        const sig = getOptionsSignature({
            address: options.address,
            chainId: ctx.chainId
        });
        if (this.instances.has(sig)) {
            console.debug(`Same address can be bind to one template only once, ignore duplicate bind: ${sig}`);
            return;
        }
        this.instances.add(sig);
        const processor = this.bindInternal({ ...options, chainId: ctx.chainId });
        for (const eh of this.logHandlers) {
            processor.onLog(eh.logIdFilter, eh.handler, eh.handlerOptions, eh.handlerName);
        }
        for (const bh of this.blockHandlers) {
            processor.onInterval(bh.handler, bh.timeIntervalInMinutes, bh.blockInterval, bh.handlerOptions, bh.handlerName);
        }
        for (const th of this.transactionHandlers) {
            processor.onTransaction(th.handler, th.handlerOptions, th.handlerName);
        }
    }
    onLog(logIdFilter, handler, handlerOptions = {}
    // fetchConfig?: Partial<FuelFetchConfig>
    ) {
        this.logHandlers.push({
            logIdFilter,
            handlerName: getHandlerName(),
            handler,
            handlerOptions
            // fetchConfig: { ...fetchConfig}
        });
        return this;
    }
    onBlockInterval(handler, blockInterval = 1000, backfillBlockInterval = 4000
    // fetchConfig?: Partial<FuelFetchConfig>
    ) {
        return this.onInterval(handler, undefined, {
            recentInterval: blockInterval,
            backfillInterval: backfillBlockInterval
        }
        // fetchConfig
        );
    }
    onTimeInterval(handler, timeIntervalInMinutes = 60, backfillBlockInterval = 240
    // fetchConfig?: Partial<FuelFetchConfig>
    ) {
        return this.onInterval(handler, { recentInterval: timeIntervalInMinutes, backfillInterval: backfillBlockInterval }, undefined
        // fetchConfig
        );
    }
    onInterval(handler, timeInterval, blockInterval, handlerOptions = {}
    // fetchConfig?: FuelFetchConfig
    ) {
        this.blockHandlers.push({
            handlerName: getHandlerName(),
            handler,
            timeIntervalInMinutes: timeInterval,
            blockInterval,
            handlerOptions
            // fetchConfig: { ...fetchConfig }
        });
        return this;
    }
    onTransaction(handler, handlerOptions = DEFAULT_FUEL_FETCH_CONFIG) {
        this.transactionHandlers.push({
            handlerName: getHandlerName(),
            handler,
            handlerOptions
        });
        return this;
    }
}
