// @typemove/iota/dist/esm/transaction.js
export function isTransactionArgument(value) {
    if (typeof value !== 'object')
        return false;
    if (value === null || value === undefined)
        return false;
    return (value.$kind === 'GasCoin' || value.$kind === 'Result' || value.$kind === 'NestedResult' || value.$kind === 'Input');
}
export function transactionArgumentOrObject(value, transactionBlock) {
    // if (isTransactionArgument(value)) {
    //   return value as TransactionArgument
    // }
    return transactionBlock.object(value);
}
export function transactionArgumentOrPure(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return typeof value == 'string' ? transactionBlock.pure.string(value) : transactionBlock.pure.u64(value);
}
export function transactionArgumentOrPureString(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.string(value);
}
export function transactionArgumentOrPureAddress(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.address(value);
}
export function transactionArgumentOrPureU8(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u8(value);
}
export function transactionArgumentOrPureU16(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u16(value);
}
export function transactionArgumentOrPureU32(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u32(value);
}
export function transactionArgumentOrPureU64(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u64(value);
}
export function transactionArgumentOrPureU128(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u128(value);
}
export function transactionArgumentOrPureU256(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.u256(value);
}
export function transactionArgumentOrPureBool(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.pure.bool(value);
}
// TODO vector might be nested
export function transactionArgumentOrVec(value, transactionBlock) {
    if (isTransactionArgument(value)) {
        return value;
    }
    return transactionBlock.makeMoveVec({
        elements: value.map((a) => transactionBlock.object(a))
    });
}
